README.md
Rush
====

A CLI tool written in pyhton for rushing web developers.

Download the latest .deb package [here](https://github.com/mnavarrocarter/rush/raw/master/dist/rush_latest_all.deb).

## Usage

This is an application in active development. No stable release has been 
published yet, so API is still undocumented.

## For Developers

To create a dev environment you need python-pip and virtualenv. 

1. Pip: `sudo install python-pip`

2. Virtualenv: `sudo pip install virtualenv`

Once you have the required packages, then you can clone the repo and 
configure your python virtual environment.

    git clone https://github.com/mnavarrocarter/rush.git Rush
    cd Rush
    virtualenv venv
    . venv/bin/activate
    pip install --editable .

Just that and you're good to go for development. Remember when you're done
to execute `deactivate`.

You can build the debian package running `./build.sh`. You will need the
following packages: 

1. `sudo apt-get install devscripts build-essential lintian`


debian/changelog
rush (0.3) unstable; urgency=low

    * Added apache user:change command
    * Set some warnings for permission changed.
    * Added suggestions

 -- Matías Navarro Carter <mnavarro@MNHPEOS>  Tue, 29 Aug 2017 10:09:00 -0300

rush (0.2) unstable; urgency=low

    * Fixed some bugs related to sudo execution

 -- Matías Navarro Carter <mnavarro@MNHPEOS>  Fri, 25 Aug 2017 16:59:08 -0300

rush (0.1) unstable; urgency=low

    * Initial release

 -- Matías Navarro Carter <mnavarro@MNHPEOS>  Wed, 23 Aug 2017 16:59:08 -0300

src/webutils.py
import click
import os

# This config object will be created and instanciated so we can share 
# global parameters and options to share with the other groups of 
# applications.
class Config(object):
    def __init__(self):
        # We define the properties here
        self.verbose = False

# We define a pass_config variable, so we pass the config.
pass_config = click.make_pass_decorator(Config, ensure=True)

###########################################################################
# WEBUTILS
###########################################################################
@click.group()
@click.option('--verbose', help="Enables verbose mode", is_flag="True")
@pass_config
def cli(config, verbose):
    """
    WebUtils by Matias Navarro Carter
    
    Version: 1.2

    This program contains some useful tools for web development.

    If you happen to have a copy of this, consider yourself lucky.

    In this current version (1.1) we only have some apache utilities related
    to Virtual Hosts management.

    You can suggest improvements or report bugs at: https://github.com/mnavarrocarter/WebUtils/issues
    """
    config.verbose = verbose

###########################################################################
# WEBUTILS->APACHE
###########################################################################
@cli.group()
@pass_config
def apache(config):
    """Apache useful commands."""
    pass

###########################################################################
# WEBUTILS->APACHE->NEWVHOST
###########################################################################
@apache.command('vhost:new', short_help='Creates a new Apache2 Virtual Host.')

@click.argument('name')

@click.option('--here', is_flag="True", 
    help="Sets the DocumentRoot to your current working directory.")
@click.option('--laravel', is_flag="True", 
    help="Adds /public to the DocumentRoot.")

@pass_config
def newvhost(config, name, here, laravel):
    """
    This command creates a new Apache2 Virtual Host

    Providing a name, it will create *.vhost-conf file in your Apache2
    virtualhost directory. We recommend the name to be something.dev like.
    Dev is a good top-level fake domain name to use for your development
    projects.

    It will create a record in your /etc/hosts file so your fake domain
    name can be resolvable in your browser.

    This script will automatically create the DocumentRoot within your 
    /var/www/vhostname.dev directory. The --here flag will set the DocumentRoot to your
    current working directory, but keep in mind that Apache will need
    permissions to access that folder. Check 'webutils apache set_perm
    --help' for more info.

    This script will require sudo privileges.
    """
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()

    if '.dev' not in name:
        if click.confirm('Do you want to add .dev to the name of your vHost? (recommended)'):
            name = name + '.dev'

    if here:
        path = os.getcwd()
        click.echo("Setting DocumentRoot to %s..." % path)
    else:
        path = '/var/www/' + name
        click.echo("Setting DocumentRoot to %s..." % path )
        if not os.path.exists('%s' % path):
            os.system('sudo mkdir %s -m 775' % path)

    if laravel:
        laravel = '/public'
        click.echo('Adding /public to the DocumentRoot for Laravel...')
    else:
        laravel = ''
    
    click.echo('Creating the .conf file...')
    out = """<VirtualHost %s:80>
    DocumentRoot %s%s
    ServerName %s
    <Directory %s%s>
        Options -Indexes +FollowSymLinks +MultiViews
        AllowOverride All
        Order allow,deny
        Allow from all
        Require all granted
    </Directory>
    ErrorLog %s/logs/error.log
    CustomLog %s/logs/access.log combined
</VirtualHost>
""" % (name, path, laravel, name, path, laravel, path, path)

    if os.path.exists('/etc/apache2/sites-available/%s.conf' % name):
        click.echo('Virtualhost already exists! Aborting...')
        quit()

    click.echo('Creating log files...')
    if not os.path.exists('%s/logs' % path):
        os.system('sudo mkdir %s/logs' % path)
    os.system('sudo touch %s/logs/error.log' % path)
    os.system('sudo touch %s/logs/access.log' % path)

    os.system('sudo echo -e "%s" >> /etc/apache2/sites-available/%s.conf' % (out, name))

    src = '/etc/apache2/sites-available/%s.conf' % name
    link = '/etc/apache2/sites-enabled/%s.conf' % name
    os.system('sudo ln -s %s %s' % (src, link))
    
    click.echo('Modifying the hosts file...')
    os.system('sudo echo -e "\n127.0.0.1 %s" >> /etc/hosts' % name)

    click.echo('Reloading Apache...')
    os.system('sudo service apache2 reload')

    click.echo('Done! Now you can visit http://%s in your browser!' % name)

###########################################################################
# WEBUTILS->APACHE->SETPERM
###########################################################################
@apache.command('perm:set', short_help='Sets special permissions for a directory.')

@click.argument('directory', default=os.getcwd())

@click.option('--apache-user', default='www-data',
    help='Overrides the default Apache User')
@click.option('--apache-group', default='www-data',
    help='Overrides the default Apache Group')
@click.option('--laravel', is_flag="True",
    help='Sets permissions for Laravel installations.')

@pass_config
def setperm(config, directory, apache_user, apache_group, laravel):
    """
    This command sets special Apache permissions for a directory.

    What this command does is that sets a directory's ownership to www-data
    user and group, so Apache can write there no matter what.
    
    If no directory is specified, then the current working directory
    is used.

    Keep in mind that you will no longer have access to that directory, because
    you are not the www-data user and you do not belong to the www-data group.
    The "webutils apache adduser" command will solve that for you, and will add your user
    to the www-data group so you can read and write those files.
    
    This command will also override your permissions in that folder,
    recursively, to 775 for folders and 664 for files.

    This command is intended to be used to give permissions to a Projects 
    folder of some sort in your home directory or alike, where you will
    have your web projects to be served.
    
    All this will require sudo privileges.
    """ 
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()

    os.system('sudo chown -R ' + apache_user + ':' + apache_group + ' ' + directory)
    os.system('sudo find ' + directory + ' -type d -exec chmod 775 {} +')
    os.system('sudo find ' + directory + ' -type d -exec chmod ug+s {} +')
    os.system('sudo find ' + directory +' -type f -exec chmod 664 {} +')
    if laravel:
        os.system('sudo chmod -R 777 ' + directory + '/storage')
        os.system('sudo chmod -R 777 ' + directory + '/bootstrap/cache')
    click.echo('Permissions applied succesfully.')
    if laravel: click.echo('Laravel special config applied.')

###########################################################################
# WEBUTILS->APACHE->UNSETPERM
###########################################################################
@apache.command('perm:unset', short_help='Unsets special permissions for a directory.')

@click.argument('directory', default=os.getcwd())

@pass_config
def unsetperm(config, directory):
    """
    This command unsets special Apache permissions for a directory.

    It will leave all folders (755) and files (664) owned by your user.
    
    If no directory is specified, then the current working directory
    is used.
    
    All this will require sudo privileges.
    """
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()

    os.system('sudo chown -R $USER:$USER ' + directory)
    os.system('sudo chmod -R 750 ' + directory)
    os.system('sudo find ' + directory +' -type f -exec chmod 660 {} +')
    click.echo('Permissions restored to default.')

###########################################################################
# WEBUTILS->APACHE->ADDUSER
###########################################################################
@apache.command('user:add', short_help='Adds the current user to the www-data group.')

@click.option('--apache-group', default='www-data',
    help='Overrides the default Apache Group')

@pass_config
def adduser(config, apache_group):
    """
    This command adds the current user to the www-data group.

    This will help you to still have access to your files while Apache controls
    them.

    You can use the --apache-group 
    """
    if os.getuid() == 0:
        click.echo('Run this as your user, duh!')
        quit()

    os.system('sudo adduser $USER ' + apache_group)

###########################################################################
# WEBUTILS->APACHE->REMUSER
###########################################################################
@apache.command('user:rem', short_help='Removes the current user from the www-data group.')

@click.option('--apache-group', default='www-data',
    help='Overrides the default Apache Group')

@pass_config
def remuser(config, apache_group):
    """
    This command removes the current user from the www-data group.

    You can use the --apache-group flag to override the www-data group.
    """ 
    if os.getuid() == 0:
        click.echo('Run this as your user, duh!')
        quit()

    os.system('sudo deluser $USER ' + apache_group)

###########################################################################
# WEBUTILS->GITHUB
###########################################################################

src/rush.py
import click
import os
import re

# This config object will be created and instanciated so we can share 
# global parameters and options to share with the other groups of 
# applications.
class Config(object):
    def __init__(self):
        # We define the properties here
        self.verbose = False

# We define a pass_config variable, so we pass the config.
pass_config = click.make_pass_decorator(Config, ensure=True)

@click.group()
@click.option('--verbose', help="Enables verbose mode", is_flag="True")
@pass_config
def cli(config, verbose):
    """
    Rush by Matias Navarro Carter
    
    Version: 0.3-beta

    This cli tool contains some useful commands for web development.

    If you happen to have a copy of this, consider yourself lucky.

    Use carefully. This is not a stable release.

    You can suggest improvements or report bugs at: https://github.com/mnavarrocarter/rush/issues
    """
    config.verbose = verbose

###########################################################################
# RUSH->NEWVHOST
###########################################################################
@cli.command('vhost:make', short_help='Creates a new Apache2 Virtual Host in your current directory.')

@click.argument('name')

@click.option('--laravel', is_flag="True", 
    help="Adds /public to the DocumentRoot.")

@pass_config
def vhostmake(config, name, laravel):
    """
    This command creates a new Apache2 Virtual Host in the current directory

    Providing a name, it will create *.vhost-conf file in your Apache2/sites-available directory. 
    We recommend the name to be something.dev like. Dev is a good top-level
    fake domain name to use for your development projects.

    Then, it will create a record in your /etc/hosts file so your fake domain
    name can be resolvable in your browser.

    Lastly, it will create the error and access logs in your project 
    directory. Make sure not to include these in your version control.

    This command requires sudo privileges.
    """
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()

    if '.dev' not in name:
        if click.confirm('Do you want to add .dev to the name of your vHost? (recommended)'):
            name = name + '.dev'

    if here:
        path = os.getcwd()
        if config.verbose:
            click.echo("Setting DocumentRoot to %s..." % path)
    else:
        path = '/var/www/' + name
        if config.verbose:
            click.echo("Setting DocumentRoot to %s..." % path )
        if not os.path.exists('%s' % path):
            os.system('sudo mkdir %s -m 775' % path)

    if laravel:
        laravel = '/public'
        if config.verbose:
            click.echo('Adding /public to the DocumentRoot for Laravel...')
    else:
        laravel = ''
    
    if config.verbose:
        click.echo('Creating the .conf file...')
    
    out = """<VirtualHost %s:80>
    DocumentRoot %s%s
    ServerName %s
    <Directory %s%s>
        Options -Indexes +FollowSymLinks +MultiViews
        AllowOverride All
        Order allow,deny
        Allow from all
        Require all granted
    </Directory>
    ErrorLog %s/logs/error.log
    CustomLog %s/logs/access.log combined
</VirtualHost>
""" % (name, path, laravel, name, path, laravel, path, path)

    if os.path.exists('/etc/apache2/sites-available/%s.conf' % name):
        click.echo('Virtualhost already exists! Aborting...')
        quit()

    if config.verbose:
        click.echo('Creating log files...')
    
    if not os.path.exists('%s/logs' % path):
        os.system('mkdir %s/logs' % path)
    os.system('touch %s/logs/error.log' % path)
    os.system('touch %s/logs/access.log' % path)
    os.system('chown 777 -R %s/logs/' %path)

    os.system('sudo echo -e "%s" >> /etc/apache2/sites-available/%s.conf' % (out, name))

    src = '/etc/apache2/sites-available/%s.conf' % name
    link = '/etc/apache2/sites-enabled/%s.conf' % name
    os.system('sudo ln -s %s %s' % (src, link))
    
    if config.verbose:
        click.echo('Modifying the hosts file...')
    os.system('sudo echo -e "\n127.0.0.1 %s" >> /etc/hosts' % name)

    if config.verbose:
        click.echo('Reloading Apache...')
    os.system('sudo service apache2 reload')

    click.echo('Done! Now you can visit http://%s in your browser!' % name)

###########################################################################
# RUSH->VHOST:DESTROY
###########################################################################
@cli.command('vhost:destroy', short_help='Destroys the given Apache2 Virtual Host.')

@click.argument('name')

@pass_config
def vhostmake(config, name):
    """
    This command destroys the given Apache2 VirtualHost. 

    This will delete the *.conf files associated with this vitualhost, as
    well as the /etc/hosts file entry, and restart Apache.

    This command requires sudo privileges.
    """
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()

    os.system('rm /etc/apache2/sites-enabled/%s.conf' % name)
    os.system('rm /etc/apache2/sites-available/%s.conf' % name)

    readFile = open('/etc/hosts')
    lines = readFile.readlines()
    readFile.close()
    for key,value in lines.items():
        if name in value:
            del lines[key]
    writeFile = open('/etc/hosts', 'w')
    writeFile.writelines(lines)
    writeFile.close()
    click.echo('Done!')

###########################################################################
# RUSH->VHOST:DEACTIVATE
###########################################################################


###########################################################################
# RUSH->VHOST:ACTIVATE
###########################################################################


###########################################################################
# RUSH->USER:CHANGE
###########################################################################
@cli.command('apacheuser:change', short_help='Sets the Apache user to the selected user.')

@click.argument('user')

@pass_config
def userchange(config, user):
    """
    This command sets the Apache user to the selected user. It requires sudo
    privileges.

    What this does is that changes apache envvars file, APACHE_RUN_USER and
    APACHE_RUN_GROUP entries, and replaces www-data for your username.

    If you want to return to the original config, just run this command again
    using www-data for the user name.
    """
    if os.getuid() != 0:
        click.echo('You need superpowers to run this thing!')
        quit()
    
    readFile = open('/etc/apache2/envvars')
    lines = readFile.readlines()
    readFile.close()
    lines[15] = "export APACHE_RUN_USER %s\n" % user
    lines[16] = "export APACHE_RUN_GROUP %s\n" % user
    w = open('/etc/apache2/envvars', 'w')
    w.writelines(lines)
    w.close()

###########################################################################
# RUSH->USER:RESTORE
###########################################################################
@cli.command('apacheuser:restore', short_help='Restores the Apache user to the www-data.')

@click.option('--no-backup', is_flag=True, help='Leaves no backup of the Apache old config.')

@pass_config
def userchange(config, no_backup):
    click.echo('Is indented dumb!')